### PR TITLE
Arnikola/m3types

### PR DESCRIPTION
**Pull request by arnikola**

### CODE REVIEW

File: models/tag.go [hunk 1]
```diff
@@ -96,8 +97,8 @@ func (m Matchers) ToTags() (Tags, error) {
 	return tags, nil
 }
 
-// ID returns a string representation of the tags
-func (t Tags) ID() string {
+// just here to prove correctness
+func (t Tags) legacyID() string {
```
Comment (nikunjgit) [inline models/tag.go]: i'd say remove it

File: models/tag.go [hunk 2]
```diff
@@ -1,14 +1,28 @@
...
+	"bytes"
 	"fmt"
 	"hash/fnv"
 	"regexp"
 	"sort"
+	"sync"
+
+	"github.com/m3db/m3x/ident"
+
+	"github.com/m3db/m3coordinator/generated/proto/prometheus/prompb"
```
Comment (benraskin92) [inline models/tag.go]: Move this above `"github.com/m3db/m3x/ident"`

File: models/tag.go [hunk 3]
```diff
@@ -85,38 +94,233 @@ type Matchers []*Matcher
...
+	return string(s)
+}
+
+type stringTags struct {
+	tags string
+}
 
-	for k := range t {
-		keys = append(keys, k)
+var _ Tags = &stringTags{}
```
Comment (nikunjgit) [inline models/tag.go]: maybe add some comments about why you're doing this ?

File: models/tag.go [hunk 4]
```diff
@@ -21,17 +35,12 @@ const (
 	MatchNotRegexp
 )
 
+var (
+	matchSymbols = [MatchNotRegexp + 1]string{"=", "!=", "=~", "!~"}
```
Comment (nikunjgit) [inline models/tag.go]: hopefully no one adds more matchtypes

File: models/tag.go [hunk 5]
```diff
@@ -85,38 +94,233 @@ type Matchers []*Matcher
 // ToTags converts Matchers to Tags
 // NB (braskin): this only works for exact matches
 func (m Matchers) ToTags() (Tags, error) {
-	tags := make(Tags, len(m))
-	for _, v := range m {
+	t := make([]*genericTag, len(m))
```
Comment (nikunjgit) [inline models/tag.go]: its a little unclear when you'd want to use genericTag vs m3Tag ?

File: models/tag.go [hunk 6]
```diff
@@ -85,38 +94,233 @@ type Matchers []*Matcher
...
+type stringID string
 
-	var b string
-	var keys []string
+func (s stringID) String() string {
+	return string(s)
+}
+
+type stringTags struct {
+	tags string
```
Comment (nikunjgit) [inline models/tag.go]: how does this work ?

File: models/tag.go [hunk 7]
```diff
@@ -85,38 +94,233 @@ type Matchers []*Matcher
...
+	identID ident.ID
+}
+
+func (id *m3ID) String() string {
+	return id.strID
+}
+
+// M3Tags is a specific tags type that optimizes for m3db backend
+type M3Tags struct {
+	tags   ident.TagIterator
```
Comment (nikunjgit) [inline models/tag.go]: when do you close ?

File: models/tag.go [hunk 8]
```diff
@@ -85,38 +94,233 @@ type Matchers []*Matcher
...
+		})
+	}
+	return labels
+}
+
+func (t *M3Tags) computeID() *m3ID {
+	return <-t.idChan
+}
+
+func (t *M3Tags) beginComputation() {
```
Comment (nikunjgit) [inline models/tag.go]: does ident not provide a way to get id ?

File: models/tag.go (hunk 4 above)
Comment (arnikola) [reply models/tag.go]: They can add them here as well ;)

File: models/tag.go (hunk 8 above)
Comment (arnikola) [reply models/tag.go]: ident does but TagIterator does not

File: models/tag.go (hunk 2 above)
Comment (arnikola) [reply models/tag.go]: If only there was a linter ;)

File: benchmark/common/parse_json.go [hunk 9]
```diff
@@ -11,6 +11,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/m3db/m3coordinator/models"
```
Comment (nikunjgit) [inline benchmark/common/parse_json.go]: group it with the next block ?

File: models/tag.go [hunk 10]
```diff
@@ -1,14 +1,29 @@
...
+
+	"github.com/m3db/m3x/ident"
 )
 
-// Tags is a key/value map of metric tags.
-type Tags map[string]string
+// Tags represents a set of metric tags
+type Tags interface {
+	ID() CoordinatorID
+	ToPromLabels() []*prompb.Label
```
Comment (nikunjgit) [inline models/tag.go]: This couples Tags with Prometheus which might be weird

File: models/genericTag/tag.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+type genericTag struct {
+	key   string
+	value string
+}
+
+// ToFormat casts this tag to the requested format.
+func (gt *genericTags) ToFormat(f models.TagFormat) (interface{}, error) {
+	if f == models.FormatProm {
```
Comment (nikunjgit) [inline models/genericTag/tag.go]: revert the check to reduce nesting

File: models/genericTag/tag.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	id   models.CoordinatorID
+}
+
+type genericTag struct {
+	key   string
+	value string
+}
+
+// ToFormat casts this tag to the requested format.
+func (gt *genericTags) ToFormat(f models.TagFormat) (interface{}, error) {
```
Comment (nikunjgit) [inline models/genericTag/tag.go]: can we have separate functions for each format ? That way you avoid the enum and also then need to type cast the interface into the format later on

File: models/interface.go [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+	// FormatProm will convert to []*prompb.Label
+	FormatProm TagFormat = iota
+	// FormatRPC will convert to []*rpc.Tags
+	FormatRPC
+)
+
+// Tags represents a set of metric tags
+type Tags interface {
+	ID() CoordinatorID
+	ToFormat(f TagFormat) (interface{}, error)
```
Comment (nikunjgit) [inline models/interface.go]: this is a bit strange to have. Also seems like there's some duplication between ToFormat in genericTags and M3Tags. How about we have ToPrometheusTags and ToRPCTags as functions which take in Tags interface ? You can then ensure Tags interface defines the functions like Name, Value ?

File: models/m3tag/tag.go [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+		t.id = t.computeID()
+	})
+	return t.id
+}
+
+// Assert types satisfy interfaces
+var _ models.Tags = &M3Tags{}
+
+// M3ID is a specific id type that optimizes for m3db backend
+func (t *M3Tags) M3ID() ident.ID {
```
Comment (nikunjgit) [inline models/m3tag/tag.go]: you have m3Id, M3ID() and ID(). I think that's a little confusing

Comment (nikunjgit) [reply models/m3tag/tag.go]: oh I see now, can you change M3ID() to call ID() and then return t.id.id ?

File: models/m3tag/tag.go [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+	t.once.Do(func() {
+		t.id = t.computeID()
+	})
+	return t.id.id
+}
+
+// GetIterator returns the tag iterator
+func (t *M3Tags) GetIterator() ident.TagIterator {
```
Comment (nikunjgit) [inline models/m3tag/tag.go]: call it TagIterator ?

File: models/m3tag/tag.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+	return ident.NewTagSliceIterator(t.tags)
+}
+
+// TODO: change id computation to use this when merged: https://github.com/m3db/m3db/pull/479
+func (t *M3Tags) beginComputation() {
```
Comment (nikunjgit) [inline models/m3tag/tag.go]: computeId or something ?

File: models/m3tag/tag.go (hunk 14 above)
Comment (nikunjgit) [reply models/m3tag/tag.go]: actually you also have a deadlock if you call both ID() and M3ID()

File: models/genericTag/tag.go [hunk 17]
```diff
@@ -0,0 +1,103 @@
+package genericTag
```
Comment (benraskin92) [inline models/genericTag/tag.go]: Rename `genericTag` in path to `generic_tag`

File: models/m3tag/tag.go [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+
+import (
+	"bytes"
+	"fmt"
+	"hash/fnv"
+	"sort"
+	"sync"
+
+	"github.com/m3db/m3coordinator/generated/proto/m3coordinator"
+
```
Comment (benraskin92) [inline models/m3tag/tag.go]: nit: remove newline

File: models/m3tag/tag.go [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+type M3Tags struct {
+	tags   ident.Tags
+	once   sync.Once
+	id     *m3ID
+	idChan chan *m3ID
+}
+
+// ToFormat casts this tag to the requested format.
+func (t *M3Tags) ToFormat(f models.TagFormat) (interface{}, error) {
+	if f == models.FormatProm {
```
Comment (benraskin92) [inline models/m3tag/tag.go]: Would using a `switch/case` statement be cleaner here? Especially if we end up adding more tag types.

File: models/matcher_test.go [hunk 20]
```diff
@@ -0,0 +1,101 @@
+package models
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
```
Comment (benraskin92) [inline models/matcher_test.go]: nit: remove newline

File: services/m3coordinator/handler/read_test.go [hunk 21]
```diff
@@ -11,6 +11,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/m3db/m3coordinator/models/m3tag"
+
```
Comment (benraskin92) [inline services/m3coordinator/handler/read_test.go]: nit: remove newline

File: storage/converter.go [hunk 22]
```diff
@@ -16,6 +18,7 @@ const (
 )
 
 // PromWriteTSToM3 converts a prometheus write query to an M3 one
+// todo benchmark
```
Comment (benraskin92) [inline storage/converter.go]: nit: put your ldap name here (e.g. `todo (arnikola): ...`

File: storage/fanout/storage_test.go [hunk 23]
```diff
@@ -6,7 +6,11 @@ import (
 	"testing"
 	"time"
 
+	"github.com/m3db/m3coordinator/models/m3tag"
+
```
Comment (benraskin92) [inline storage/fanout/storage_test.go]: nit: remove newline

File: storage/local/storage.go [hunk 24]
```diff
@@ -149,16 +152,21 @@ func (s *localStorage) Type() storage.Type {
...
 	store := common.store
-	id := ident.StringID(common.id)
-	namespace := ident.StringID(store.namespace)
-	return store.session.Write(namespace, id, w.timestamp, w.value, common.unit, common.annotation)
+	tags, ok := common.tags.(*m3tag.M3Tags)
+	if !ok {
+		id := ident.StringID(common.tags.ID().String())
+		return store.session.Write(store.namespace, id, w.timestamp, w.value, common.unit, common.annotation)
+	}
+	// it := tags.GetIterator()
```
Comment (benraskin92) [inline storage/local/storage.go]: ??

File: tsdb/remote/codecs.go [hunk 25]
```diff
@@ -4,6 +4,9 @@ import (
 	"context"
 	"time"
 
+	"github.com/m3db/m3coordinator/models/m3tag"
+
```
Comment (benraskin92) [inline tsdb/remote/codecs.go]: nit: remove newline

File: tsdb/remote/codecs_test.go [hunk 26]
```diff
@@ -5,6 +5,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/m3db/m3coordinator/models/m3tag"
+
```
Comment (benraskin92) [inline tsdb/remote/codecs_test.go]: nit: remove newline

File: tsdb/remote/codecs_test.go [hunk 27]
```diff
@@ -30,14 +32,24 @@ var (
 	mps1     = int32(120)
 	time1    = "2093-02-06T11:54:48+07:00"
 
-	tags0  = map[string]string{"a": "b", "c": "d"}
-	tags1  = map[string]string{"e": "f", "g": "h"}
+	tags0  = generateTags("abc", "def")
```
Comment (benraskin92) [inline tsdb/remote/codecs_test.go]: nit: stick to `foo`, `bar`, `biz`, `baz`, etc.

File: tsdb/remote/server_test.go [hunk 28]
```diff
@@ -8,8 +8,9 @@ import (
 	"testing"
 	"time"
 
+	"github.com/m3db/m3coordinator/models/m3tag"
+
```
Comment (benraskin92) [inline tsdb/remote/server_test.go]: nit: remove newline

File: tsdb/remote/server_test.go [hunk 29]
```diff
@@ -26,7 +27,7 @@ const (
 
 var (
 	startTime, _ = time.Parse(time.RFC3339, "2000-02-06T11:54:48+07:00")
-	tags         = models.Tags{"1": "b", "2": "c"}
+	tags         = generateTags("abc", "def")
```
Comment (benraskin92) [inline tsdb/remote/server_test.go]: same here

File: tsdb/remote/codecs_test.go [hunk 30]
```diff
@@ -30,14 +32,24 @@ var (
...
+	tags1  = generateTags("ghi", "jkl")
 	float0 = 100.0
 	float1 = 3.5
 	ann    = []byte("aasjgał")
 	id     = "asdgsdh"
 )
 
+func generateTags(name, value string) *rpc.Tags {
+	tags := []*rpc.Tag{
+		&rpc.Tag{
```
Comment (benraskin92) [inline tsdb/remote/codecs_test.go]: I think you can rewrite like:

```
tags := []*rpc.Tag{{
   Name: name,
   Value: value,
  },
}
```

Comment (benraskin92) [reply tsdb/remote/codecs_test.go]: Couple other places where you can do a similar thing.

Comment (arnikola) [reply tsdb/remote/codecs_test.go]: Nice!

File: models/interface.go [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+			Name:  tag.Key,
+			Value: tag.Value,
+		})
+	}
+
+	return labels
+}
+
+// TagsToRPCTags converts a list of tags to prometheus labels
+func TagsToRPCTags(t Tags) *rpc.Tags {
```
Comment (nikunjgit) [inline models/interface.go]: better naming RPC can mean anything

File: models/m3tag/tag.go [hunk 32]
```diff
@@ -0,0 +1,170 @@
...
+// Assert types satisfy interfaces
+var _ models.Tags = &M3Tags{}
+
+// TagIterator returns a tag iterator for these tags
+func (t *M3Tags) TagIterator() ident.TagIterator {
+	return ident.NewTagSliceIterator(t.tags)
+}
+
+// Finalize releases the internal ident.Tags to their object pool
+func (t *M3Tags) Finalize() {
```
Comment (nikunjgit) [inline models/m3tag/tag.go]: remove the prints ?

File: models/m3tag/tag.go [hunk 33]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+// TagIteratorToM3Tags wraps a TagIterator into an M3Tags
+func TagIteratorToM3Tags(it ident.TagIterator) *M3Tags {
+	t := make([]ident.Tag, 0, it.Remaining())
+
+	for it.Next() {
+		t = append(t, it.Current())
+	}
+	it.Close()
```
Comment (nikunjgit) [inline models/m3tag/tag.go]: add it as a defer on the first line of function ?

File: models/m3tag/tag.go [hunk 34]
```diff
@@ -0,0 +1,170 @@
...
+	t := make([]ident.Tag, 0, len(tags))
+
+	for _, tag := range tags {
+		t = append(t, ident.StringTag(tag.GetName(), tag.GetValue()))
+	}
+
+	return createM3Tags(t)
+}
+
+type ascendingByKeyStringProm []*prompb.Label
```
Comment (nikunjgit) [inline models/m3tag/tag.go]: would it be better to define []*prompb.Label as a type and use it everywhere ? You can define these functions on that type

File: models/m3tag/tag.go [hunk 35]
```diff
@@ -0,0 +1,170 @@
...
+	sort.Sort(ascendingByKeyStringProm(labels))
+
+	for _, label := range labels {
+		t = append(t, ident.StringTag(label.Name, label.Value))
+	}
+
+	return createM3Tags(t)
+}
+
+type ascendingByKeyMatcher models.Matchers
```
Comment (nikunjgit) [inline models/m3tag/tag.go]: why not just define these functions on models.Matchers

File: storage/fanout/storage.go [hunk 36]
```diff
@@ -83,6 +85,13 @@ func (s *fanoutStorage) Write(ctx context.Context, query *storage.WriteQuery) er
 		requests[idx] = newWriteRequest(store, query)
 	}
 
+	if query != nil && query.Tags != nil {
+		if m3tags, ok := query.Tags.(*m3tag.M3Tags); ok {
+			// Close internal tag iterators
+			defer m3tags.Finalize()
```
Comment (nikunjgit) [inline storage/fanout/storage.go]: this is a little strange. Maybe define Finalize() on the interface ?

Comment (arnikola) [reply storage/fanout/storage.go]: Sounds good; I guess not all tags need to be finalizeable, but they can just implement a noop

File: models/interface.go [hunk 37]
```diff
@@ -0,0 +1,66 @@
...
+		})
+	}
+
+	return &rpc.Tags{
+		Tags: tags,
+	}
+}
+
+// PrometheusLabels is a list of prometheus labels
+type PrometheusLabels []*prompb.Label
```
Comment (nikunjgit) [inline models/interface.go]: PromLabels ?

**Comment by arnikola** [comment]
Closing this as it doesn't align with future work, and did not lead to significant performance improvements.

**Pull request closed without merging**